### ISSUE TITLE
[Flaky test] [sig-storage] [Serial] Volume metrics PVController should create bound pv/pvc count metrics for pvc controller after creating both pv and pvc

### ISSUE THREAD

**Issue by mariantalla** [kubernetes/kubernetes#75275]
**Which jobs are flaking**:
* [ci-kubernetes-e2e-gce-new-master-upgrade-master](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/2312) ([testgrid](https://testgrid.k8s.io/sig-release-master-upgrade#gce-new-master-upgrade-master
))
* [ci-kubernetes-e2e-gce-new-master-upgrade-cluster-new/2154](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-cluster-new/2154) ([testgrid](https://testgrid.k8s.io/sig-release-master-upgrade#gce-new-master-upgrade-cluster-new))

**Which test(s) are flaking**:
`[sig-storage] [Serial] Volume metrics PVController should create bound pv/pvc count metrics for pvc controller after creating both pv and pvc`

**Reason for failure**:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/volume_metrics.go:412
Wrong pv controller metric pv_collector_bound_pvc_count(namespace): wanted map[pv-7445:1], got map[]
Expected
    <map[string]int64 | len:0>: {}
to equal
    <map[string]int64 | len:1>: {"pv-7445": 1}
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/volume_metrics.go:355
```

**Anything else we need to know**:
* At the moment, this test fails anywhere from 25% to 40% of the time in the above release blocking jobs.
* [Triage](https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Volume%20metrics%20PVController%20should%20create%20bound%20pv%2Fpvc%20count%20metrics%20for%20pvc%20controller%20after%20creating%20both%20pv%20and%20pvc)

/sig storage
/kind flake
/remove-kind failing-test
/priority important-soon
/milestone v1.14

cc @kubernetes/sig-storage-test-failures 
cc @smourapina @alejandrox1 @kacole2 @mortent

**Comment by gnufied** [comment]
@bertinatto has volunteered to work on this. Assigning it to him.

/assign @bertinatto

**Comment by gnufied** [comment]
Fabio has volunteered to work on this. 

/assign @bertinatto

**Comment by mariantalla** [comment]
Hey @bertinatto @gnufied just checking in on this. Any findings?

**Comment by bertinatto** [comment]
I've run this test in a loop for around 1.5 hours but hasn't flaked.

This is what I think is happening:

1. PV Controller is running and has some not-yet-syncronized metrics in-memory. The standard interval that metrics are collected is 1m: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#configuration-file
2. PV Controller is upgraded and restarted
3. The running test checks for the PV controller metrics and finds nothing there

**Comment by mariantalla** [comment]
Thanks @bertinatto . Is this expected? Is the PV controller expected to have persisted any in-memory metrics before shutting down? Or is it best-effort?

**Comment by bertinatto** [comment]
Unfortunately so far I couldn't neither reproduce the flake nor find any proof that my theory is correct.

@mlmhl, since you created that test, do you have any thoughts on why this fake is happening?

PS: My suspicion comes from the fact that the flake only happens in `*upgrade` jobs, but I couldn't find any proof that it's correct.

**Comment by msau42** [comment]
This test is not run *across* an upgrade.

The upgrade happens and finishes, and then the test will run.

**Comment by mlmhl** [comment]
Hi @bertinatto 

> Wrong pv controller metric pv_collector_bound_pvc_count(namespace): wanted map[pv-7445:1], got map[]

From the error message, we can see that the bound pvc number is incorrect. However, in the test case, we check the bound pv number first before checking bound pvc number, so this means the bound pv number is correct, but bound pvc number is incorrect. This is odd.

Can anyone help to post more details of the upgrade job?

**Comment by bertinatto** [comment]
@mlmhl, I think we do get at least 1 PV/PVC when we wait [here](https://github.com/kubernetes/kubernetes/blob/897d62ace79a8fa72f9825e065b2b67fdba65bec/test/e2e/storage/volume_metrics.go#L417-L418), however, the test fails [here](https://github.com/kubernetes/kubernetes/blob/897d62ace79a8fa72f9825e065b2b67fdba65bec/test/e2e/storage/volume_metrics.go#L353-L358) when we check the number of PVCs again. So we check once and it succeeds, we check twice and it fails. That's why I suspected the controller was reloading, but as @msau42 pointed out that's not the case.

Do you think `calculateRelativeValues` could be misbehaving depending on the input that it's given?

**Comment by mlmhl** [comment]
> I think we do get at least 1 PV/PVC when we wait here, however, the test fails here when we check the number of PVCs again.

Yeah, I agree with this. So I wonder if we have multi jobs running in parallel? For example:

1. Before we starting this test, some other jobs created a PVC `A`, and the original bound PVC number would be 1
2. Then this test started and created a PVC `B` too, meanwhile the other job deleted the PVC `A`, so the updated PVC number is till 1, and the `calculateRelativeValues` method will return 0.

**Comment by bertinatto** [comment]
> Yeah, I agree with this. So I wonder if we have multi jobs running in parallel? For example:

@mlmhl, since this in the `[Serial]` test suit there shouldn't be other jobs running, no?

**Comment by bertinatto** [comment]
We do run serial and [disruptive tests along with serial:](https://github.com/kubernetes/test-infra/blob/f9d36b8fe58dc85f6f2c59106cca64f529ff06ac/config/jobs/kubernetes/sig-cluster-lifecycle/k8s-upgrade-gce.yaml#L544). Looks like this change is [fairly recent](https://github.com/kubernetes/test-infra/commit/29362201b30f4f8b8d7a3c0dfa68d67915447b88).

@msau42, do you think this makes sense?

**Comment by msau42** [comment]
I think running serial and disruptive together is fine. We do that for non-upgrade jobs too.

**Comment by bertinatto** [comment]
This job from March 11:

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/2312/build-log.txt

This is where it fails:

```
I0311 16:58:35.177] 
I0311 16:58:35.177]     [91mWrong pv controller metric pv_collector_bound_pvc_count(namespace): wanted map[e2e-tests-pv-zrz6h:1], got map[]
I0311 16:58:35.177]     Expected
I0311 16:58:35.177]         <map[string]int64 | len:0>: {}
I0311 16:58:35.178]     to equal
I0311 16:58:35.178]         <map[string]int64 | len:1>: {
I0311 16:58:35.178]             "e2e-tests-pv-zrz6h": 1,
I0311 16:58:35.178]         }[0m
I0311 16:58:35.178] 
```

Around 10 minutes earlier, it runs this test:

```
I0311 16:58:23.604]   Local volume provisioner [Serial]
I0311 16:58:23.605]   [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/persistent_volumes-local.go:433[0m
I0311 16:58:23.605]     should create and recreate local persistent volume
```

Which was [removed in January 2019](https://github.com/kubernetes/kubernetes/pull/72617/files#diff-f87246d95872c9c1ea417fd6419dc6e1L417).

So I think this job is running tests from an old tree.

**Comment by msau42** [comment]
The test only upgrades the master, and runs the 1.13 version of the test: 

I0311 15:08:44.679] Mar 11 15:08:44.679: INFO: e2e test version: v1.13.5-beta.0.7+6c1e64b94a3e11

**Comment by msau42** [comment]
Are we sure that no other tests were running at the same time?  If you search pv controller logs, do you see other pvs being created within the time of this test?

**Comment by bertinatto** [comment]
None of the logs I saw had other tests running at the same timeframe, but I'll check again.

**Comment by bertinatto** [comment]
This flake indeed seems to be caused by a failing upgrade. From the logs:

```
W0311 15:08:26.309] 2019/03/11 15:08:26 main.go:307: Something went wrong: encountered 1 errors: [error during ./hack/ginkgo-e2e.sh --ginkgo.focus=\[Feature:MasterUpgrade\] --upgrade-target=ci/latest --report-dir=/workspace/_artifacts --disable-log-dump=true --report-prefix=upgrade: exit status 1]
```

It hasn't happened in almost 3 days, so I I think [this workaround](https://github.com/kubernetes/kubernetes/pull/75305), [later reverted](https://github.com/kubernetes/kubernetes/pull/75393), worked. The proper fix would be the [go 1.12.1 update](https://github.com/kubernetes/kubernetes/issues/75372), which happened recently.

I'll keep an eye on it, but it seems to be resolved.

**Comment by spiffxp** [comment]
Yeah this seems like it has been addressed by the move to go 1.12.1, moving to observation

https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Volume%20metrics%20PVController%20should%20create%20bound%20pv%2Fpvc%20count%20metrics%20for%20pvc%20controller%20after%20creating%20both%20pv%20and%20pvc
<img width="1101" alt="Screen Shot 2019-03-16 at 12 50 33 PM" src="https://user-images.githubusercontent.com/49258/54480794-1c0c5380-47ea-11e9-9379-32045e06398a.png">

**Comment by spiffxp** [comment]
/close
Situation unchanged, calling this resolved
